src/sleep/sleep.c

```diff
@@ -240,6 +240,52 @@ static int lock_all_homes(void) {
         return 0;
 }
 
+/*
+ * The 'wakeup_count' attribute provides a means by which wakeup events can be
+ * handled in a non-racy way.
+ *
+ * If a wakeup event occurs when the system is in a sleep state, it simply is
+ * woken up.  In turn, if an event that would wake the system up from a sleep
+ * state occurs when it is undergoing a transition to that sleep state, the
+ * transition should be aborted.  Moreover, if such an event occurs when the
+ * system is in the working state, an attempt to start a transition to the
+ * given sleep state should fail during certain period after the detection of
+ * the event.  Using the 'state' attribute alone is not sufficient to satisfy
+ * these requirements, because a wakeup event may occur exactly when 'state'
+ * is being written to and may be delivered to user space right before it is
+ * frozen, so the event will remain only partially processed until the system is
+ * woken up by another event.  In particular, it won't cause the transition to
+ * a sleep state to be aborted.
+ *
+ * For more details see https://lwn.net/Articles/392897/
+ */
+static int write_wakeup_count(void) {
+        static const char kernel_wakeup_count[] = "/sys/power/wakeup_count";
+        static const char systemd_wakeup_count[] = "/run/wakeup_count";
+        _cleanup_free_ char *buf = NULL;
+        size_t bufsize;
+        int r;
+
+        if (!access(systemd_wakeup_count, F_OK)) {
+                r = read_full_file(systemd_wakeup_count, &buf, &bufsize);
+                if (r < 0)
+                        return log_debug_errno(r, "Unable to read %s: %m", systemd_wakeup_count);
+                r = unlink(systemd_wakeup_count);
+                if (r < 0)
+                        return log_debug_errno(r, "Unable to unlink %s: %m", systemd_wakeup_count);
+        } else {
+                r = read_full_virtual_file(kernel_wakeup_count, &buf, &bufsize);
+                if (r < 0)
+                        return log_debug_errno(r, "Unable to read %s: %m", kernel_wakeup_count);
+        }
+
+        r = write_string_file(kernel_wakeup_count, buf, WRITE_STRING_FILE_DISABLE_BUFFER);
```

Comment:
so this will fail if a wakeup was seen in the meantime. we need to look for that error code and then print a nice info log message, and then return some clear return value, that makes sure the sleep stuff exits early.